File: src/database/ReconciliationRunRepository.cpp
#include "database/ReconciliationRunRepository.hpp"
#include "database/DatabaseException.hpp"

#include <pqxx/pqxx>

namespace reconciliation::database {

ReconciliationRunRepository::ReconciliationRunRepository(DatabaseConnection& connection)
    : connection_(connection) {}

int64_t ReconciliationRunRepository::createRun(const std::string& algorithm,
                                                const std::string& matchingStrategy,
                                                int64_t amountToleranceCents) {
    try {
        pqxx::work txn(connection_.handle());
        pqxx::result r = txn.exec_params(
            "INSERT INTO reconciliation_runs (algorithm, matching_strategy, amount_tolerance_cents) "
            "VALUES ($1, $2, $3) RETURNING id",
            algorithm, matchingStrategy, amountToleranceCents
        );
        txn.commit();
        return r[0][0].as<int64_t>();
    } catch (const pqxx::sql_error& e) {
        throw DatabaseException(std::string("createRun failed: ") + e.what());
    }
}

void ReconciliationRunRepository::completeRun(int64_t runId, const RunStats& stats, int64_t durationMs) {
    try {
        pqxx::work txn(connection_.handle());
        txn.exec_params(
            "UPDATE reconciliation_runs SET "
            "  records_internal = $1, records_external = $2, matched_count = $3, "
            "  mismatched_count = $4, missing_internal_count = $5, missing_external_count = $6, "
            "  duplicate_count = $7, duration_ms = $8, completed_at = now() "
            "WHERE id = $9",
            stats.recordsInternal, stats.recordsExternal, stats.matchedCount,
            stats.mismatchedCount, stats.missingInternalCount, stats.missingExternalCount,
            stats.duplicateCount, durationMs, runId
        );
        txn.commit();
    } catch (const pqxx::sql_error& e) {
        throw DatabaseException(std::string("completeRun failed: ") + e.what());
    }
}

} // namespace reconciliation::database

File: recon_engine/database/reconciliation_run_repository.py
from __future__ import annotations

from dataclasses import dataclass

import psycopg2

from .connection import DatabaseConnection, DatabaseException


@dataclass
class RunStats:
    records_internal: int = 0
    records_external: int = 0
    matched_count: int = 0
    mismatched_count: int = 0  # sum of AMOUNT/TAX/DATE/VENDOR/MULTIPLE mismatch statuses
    missing_internal_count: int = 0
    missing_external_count: int = 0
    duplicate_count: int = 0


class ReconciliationRunRepository:
    def __init__(self, connection: DatabaseConnection):
        self._connection = connection

    def create_run(self, algorithm: str, matching_strategy: str, amount_tolerance_cents: int) -> int:
        try:
            with self._connection.handle.cursor() as cur:
                cur.execute(
                    "INSERT INTO reconciliation_runs (algorithm, matching_strategy, amount_tolerance_cents) "
                    "VALUES (%s, %s, %s) RETURNING id",
                    (algorithm, matching_strategy, amount_tolerance_cents),
                )
                run_id = cur.fetchone()[0]
            self._connection.handle.commit()
            return run_id
        except psycopg2.Error as e:
            self._connection.handle.rollback()
            raise DatabaseException(f"createRun failed: {e}") from e

    def complete_run(self, run_id: int, stats: RunStats, duration_ms: int) -> None:
        try:
            with self._connection.handle.cursor() as cur:
                cur.execute(
                    "UPDATE reconciliation_runs SET "
                    "  records_internal = %s, records_external = %s, matched_count = %s, "
                    "  mismatched_count = %s, missing_internal_count = %s, missing_external_count = %s, "
                    "  duplicate_count = %s, duration_ms = %s, completed_at = now() "
                    "WHERE id = %s",
                    (
                        stats.records_internal, stats.records_external, stats.matched_count,
                        stats.mismatched_count, stats.missing_internal_count, stats.missing_external_count,
                        stats.duplicate_count, duration_ms, run_id,
                    ),
                )
            self._connection.handle.commit()
        except psycopg2.Error as e:
            self._connection.handle.rollback()
            raise DatabaseException(f"completeRun failed: {e}") from e
